File: src/mulinex_ignition/src/move_wheels.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"
#include "pi3hat_moteus_int_msgs/msg/omni_mulinex_command.hpp"
 
using namespace std::chrono_literals;
 
int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("omni_vel_commander");
 
  // parameters
  node->declare_parameter<double>("vx", 0.5);
  node->declare_parameter<double>("vy", 0.0);
  node->declare_parameter<double>("w", 0.0);
  node->declare_parameter<int>("publish_rate", 10);
  double vx = node->get_parameter("vx").as_double();
  double vy = node->get_parameter("vy").as_double();
  double w  = node->get_parameter("w").as_double();
  int rate_hz = node->get_parameter("publish_rate").as_int();
 
  // QoS (match BestEffort + deadline)
  rclcpp::QoS qos(10);
  qos.reliability(rclcpp::ReliabilityPolicy::BestEffort);
  qos.deadline(std::chrono::milliseconds(1000 / rate_hz + 5));
 
  // 1) publish the correct message type
  using CmdMsg = pi3hat_moteus_int_msgs::msg::OmniMulinexCommand;
  // 2) on the same topic the joystick uses
  auto pub = node->create_publisher<CmdMsg>("/omni_control/command", qos);
 
  // --- HOMING (must set data=true) ---
  using HomingSvc = std_srvs::srv::SetBool;
  auto homing_cli =
    node->create_client<HomingSvc>("/omni_control/homing_srv");
  if (!homing_cli->wait_for_service(2s)) {
    RCLCPP_ERROR(node->get_logger(), "Homing service not available");
    return 1;
  }
  auto homing_req = std::make_shared<HomingSvc::Request>();
  homing_req->data = true;
  auto homing_fut = homing_cli->async_send_request(homing_req);
  if (rclcpp::spin_until_future_complete(node, homing_fut, 2s)
        != rclcpp::FutureReturnCode::SUCCESS
      || !homing_fut.get()->success)
  {
    RCLCPP_ERROR(node->get_logger(), "Homing failed");
    return 1;
  }
  RCLCPP_INFO(node->get_logger(), "Homing succeeded—controller ACTIVE");
 
  // record start time to enforce timeout
  rclcpp::Time start_time = node->now();
  rclcpp::Duration timeout_duration(5s);
 
  // publish loop
  rclcpp::Rate loop_rate(rate_hz);
  while (rclcpp::ok()) {
    auto msg = CmdMsg();
    // after 5 seconds, override velocities to zero
    bool timeout = (node->now() - start_time) > timeout_duration;
    msg.set__v_x(timeout ? 0.0 : vx);
    msg.set__v_y(timeout ? 0.0 : vy);
    msg.set__omega(timeout ? 0.0 : w);
    msg.set__height_rate(0.0);                      // joystick always sets this
    msg.header.stamp = node->now();                 // stamped header
 
    pub->publish(msg);
    rclcpp::spin_some(node);
    loop_rate.sleep();
  }
 
  rclcpp::shutdown();
  return 0;
}

File: src/mulinex_ignition_py/mulinex_ignition_py/move_wheels.py
import rclpy
from rclpy.node import Node
from std_srvs.srv import SetBool
from pi3hat_moteus_int_msgs.msg import OmniMulinexCommand
from rclpy.qos import QoSProfile, ReliabilityPolicy
from rclpy.duration import Duration
from rclpy.time import Time


''' This node moves the robot using the omni_velocity controller.
This controller is used to control the robot's movement in a 2D plane. 
It takes linear and angular velocities as input, solves the inverse kinematics, and sends commands to the motors.'''

class OmniVelCommander(Node):
    def __init__(self):
        super().__init__('omni_vel_commander')

        # Declare parameters with default values
        self.declare_parameter('vx', 0.5)
        self.declare_parameter('vy', -0.0)
        self.declare_parameter('w', -0.5)
        self.declare_parameter('publish_rate', 10)

        # Get parameter values
        self.vx = self.get_parameter('vx').get_parameter_value().double_value
        self.vy = self.get_parameter('vy').get_parameter_value().double_value
        self.w  = self.get_parameter('w').get_parameter_value().double_value
        self.rate_hz = self.get_parameter('publish_rate').get_parameter_value().integer_value

        # QoS profile
        qos = QoSProfile(
            depth=10,
            reliability=ReliabilityPolicy.BEST_EFFORT,
            #deadline=QoSDuration(nanoseconds=int(1e9 / self.rate_hz + 5e6))
            deadline = Duration(seconds=0, nanoseconds=int(1e9 / self.rate_hz + 5e6))

        )
        
        # Publisher
        self.publisher = self.create_publisher(OmniMulinexCommand, '/omni_control/command', qos)

        # Homing service
        self.homing_client = self.create_client(SetBool, '/omni_control/homing_srv')
        if not self.homing_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().error('Homing service not available')
            rclpy.shutdown()
            return

        req = SetBool.Request()
        req.data = True

        future = self.homing_client.call_async(req)
        rclpy.spin_until_future_complete(self, future, timeout_sec=2.0)
        if not future.result() or not future.result().success:
            self.get_logger().error('Homing failed')
            rclpy.shutdown()
            return

        self.get_logger().info('Homing succeeded—controller ACTIVE')

        # Timer and time tracking
        self.start_time = self.get_clock().now()
        self.timeout_duration = Duration(seconds=10.0)

        self.timer = self.create_timer(1.0 / self.rate_hz, self.publish_cmd)

    def publish_cmd(self):
        now = self.get_clock().now()
        timeout = (now - self.start_time) > self.timeout_duration

        msg = OmniMulinexCommand()
        msg.v_x = 0.0 if timeout else self.vx
        msg.v_y = 0.0 if timeout else self.vy
        msg.omega = 0.0 if timeout else self.w
        msg.height_rate = 0.0
        msg.header.stamp = now.to_msg()

        self.publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = OmniVelCommander()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()
